#[wasm_bindgen] attribute not work in macro_rules

### Describe the Bug

When putting struct attributed `#[wasm_bindgen]` in a custom macro_rules, it throws error `cannot find value 'js' in this scope`.

### Steps to Reproduce

code snippet:

```rust
// a macro that output exactly the same input struct
macro_rules! make_struct {
    {
        $( #[$attr:meta] )*
        $svis:vis struct $record_name:ident {
            $(
                $fvis:vis $fname:ident : $ftype:ty
            ),* $(,)?
        }
    } => {
        $( #[$attr] )*
        $svis struct $record_name {
            $(
                $fvis $fname : $ftype
            ),*
        }
    };
}

// testing
make_struct!{
    #[wasm_bindgen]
    #[derive(Debug, Clone, Copy)]
    pub struct Test {
        pub a: i32 // error: cannot find value `js` in this scope
    }
}
```

### Expected Behavior

In my imagination, macro rules should be expanded before the proc macros, thus the above code should work as normal.

Correct me if I'm wrong.

Hello, I'm also experiencing this issue.
I've created a simple reproducible repository here: https://github.com/ChiaMineJP/error-report/tree/main/wasm-bindgen-with-macro

# Summary
I defined a simple macro rule as below, which just adds a line `#[wasm-bindgen]` to
a struct definition.

```rust
macro_rules! with_wasm {
    ($name:ident {$field:ident: $t:ty}) => {
        #[wasm_bindgen::prelude::wasm_bindgen]
        pub struct $name {
            pub $field: $t
        }
    }
}
```

This macro simply converts
```rust
with_wasm!(AStruct {
    a_field: u8
});
```
to
```rust
#[wasm_bindgen::prelude::wasm_bindgen]
pub struct AStruct {
    pub a_field: u8,
}
```

While the expanded code (the below) passes `wasm-pack build` without error,
the original code with macro (the above) fails with error:
```
error[E0425]: cannot find value `js` in this scope
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,11 @@
 * Add bindings for `UserActivation`.
   [#3719](https://github.com/rustwasm/wasm-bindgen/pull/3719)
 
+### Fixed
+
+* Fixed a compiler error when using `#[wasm_bindgen]` inside `macro_rules!`.
+  [#3725](https://github.com/rustwasm/wasm-bindgen/pull/3725)
+
 ### Removed
 
 * Removed Gecko-only `InstallTriggerData` and Gecko-internal `FlexLineGrowthState`, `GridDeclaration`, `GridTrackState`,

--- a/crates/backend/src/codegen.rs
+++ b/crates/backend/src/codegen.rs
@@ -389,7 +389,14 @@ impl ToTokens for ast::StructField {
         };
         let maybe_assert_copy = respan(maybe_assert_copy, ty);
 
-        let mut val = quote! { (*js).borrow().#rust_name };
+        // Split this out so that it isn't affected by `quote_spanned!`.
+        //
+        // If we don't do this, it might end up being unable to reference `js`
+        // properly because it doesn't have the same span.
+        //
+        // See https://github.com/rustwasm/wasm-bindgen/pull/3725.
+        let js_token = quote! { js };
+        let mut val = quote_spanned!(self.rust_name.span()=> (*#js_token).borrow().#rust_name);
         if let Some(span) = self.getter_with_clone {
             val = quote_spanned!(span=> <#ty as Clone>::clone(&#val) );
         }

--- a/crates/backend/src/codegen.rs
+++ b/crates/backend/src/codegen.rs
@@ -389,7 +389,7 @@ impl ToTokens for ast::StructField {
         };
         let maybe_assert_copy = respan(maybe_assert_copy, ty);
 
-        let mut val = quote_spanned!(self.rust_name.span()=> (*js).borrow().#rust_name);
+        let mut val = quote! { (*js).borrow().#rust_name };
         if let Some(span) = self.getter_with_clone {
             val = quote_spanned!(span=> <#ty as Clone>::clone(&#val) );
         }

--- a/tests/wasm/classes.js
+++ b/tests/wasm/classes.js
@@ -241,3 +241,10 @@ exports.js_test_inspectable_classes_can_override_generated_methods = () => {
     assert.strictEqual(overridden_inspectable.toString(), 'string was overwritten');
     overridden_inspectable.free();
 };
+
+exports.js_test_class_defined_in_macro = () => {
+    const macroClass = new wasm.InsideMacro();
+    assert.strictEqual(macroClass.a, 3);
+    macroClass.a = 5;
+    assert.strictEqual(macroClass.a, 5);
+};

--- a/tests/wasm/classes.rs
+++ b/tests/wasm/classes.rs
@@ -34,6 +34,7 @@ extern "C" {
     fn js_test_option_classes();
     fn js_test_inspectable_classes();
     fn js_test_inspectable_classes_can_override_generated_methods();
+    fn js_test_class_defined_in_macro();
 }
 
 #[wasm_bindgen_test]
@@ -607,4 +608,28 @@ impl OverriddenInspectable {
     pub fn js_to_string(&self) -> String {
         String::from("string was overwritten")
     }
+}
+
+macro_rules! make_struct {
+    ($field:ident) => {
+        #[wasm_bindgen]
+        pub struct InsideMacro {
+            pub $field: u32,
+        }
+    };
+}
+
+make_struct!(a);
+
+#[wasm_bindgen]
+impl InsideMacro {
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self { a: 3 }
+    }
+}
+
+#[wasm_bindgen_test]
+fn class_defined_in_macro() {
+    js_test_class_defined_in_macro();
 }

